runtime/src/main/java/org/corfudb/runtime/collections/StreamPollingTask.java

```diff
@@ -0,0 +1,108 @@
+ * <p>
+ * Created by WenbinZhu on 11/9/20.
+ */
+@Slf4j
+class StreamPollingTask implements Runnable {
+
+    // A period of time in ms to sleep before next cycle when poller gets no new data changes.
+    private static final Duration IDLE_WAIT_TIME_MS = Duration.ofMillis(50);
+
+    // Total amount of time to wait for putting the polled data changes into buffer if it is full.
+    private static final Duration QUEUE_FULL_BLOCK_TIME_MS = Duration.ofMillis(2_000);
+
+    // The streaming manager that is in charge of listener subscriptions.
+    private final StreamingManager streamingManager;
+
+    // The corfu transaction stream to poll data changes from.
+    private final IStreamView txnStream;
+
+    // The subscription context associated with this task.
+    private final StreamSubscription subscription;
+
+    // The Thread pool for executing stream polling tasks.
+    private final ExecutorService pollingExecutor;
+
+    // Last address of the data successfully processed by the buffer.
+    private long lastReadAddress;
+
+    StreamPollingTask(StreamingManager streamingManager, long lastAddress,
+                      StreamSubscription subscription, ExecutorService executor) {
+        this.streamingManager = streamingManager;
+        this.subscription = subscription;
+        this.pollingExecutor = executor;
+        this.lastReadAddress = lastAddress;
+        this.txnStream = subscription.getTxnStream();
+    }
+
+    @Override
+    public void run() {
+        try {
+            pollTxnStream();
+        } catch (Throwable throwable) {
+            StreamListener listener = subscription.getListener();
+            log.error("Encountered exception {} during txn stream polling, listener: {}, " +
+                    "namespace: {}", throwable, listener, subscription.getNamespace());
+            streamingManager.unsubscribe(listener, false);
+            listener.onError(throwable);
+        }
+    }
+
+    /**
+     * Poll new data changes from the transaction stream and put into
+     * the subscription's buffer one by one.
+     */
+    private void pollTxnStream() throws Exception {
+        // If listener already unsubscribed, do not process or schedule again.
+        if (subscription.isStopped()) {
+            return;
+        }
+
+        // Seek to next address and poll transaction updates.
+        txnStream.seek(lastReadAddress + 1L);
+        List<ILogData> updates = txnStream.remaining();
+
+        // No new updates, take a short break and poll again.
+        if (updates.isEmpty()) {
+            TimeUnit.MILLISECONDS.sleep(IDLE_WAIT_TIME_MS.toMillis());
+            pollingExecutor.submit(this);
+            return;
+        }
+
+        // Insert polled updates to the subscription buffer, with a shared
+        // fixed amount of time waiting for buffer being not full.
+        long remainingBlockTime = QUEUE_FULL_BLOCK_TIME_MS.toNanos();
+        for (ILogData update : updates) {
```

Comment:
If the goal of using fixed size queue and retry was to avoid starvation of other streams when the number of subscribers exceeds the number of threads. Then won't this still bring back the starvation if the number of updates are large?

Reply:
This is why the QUEUE_FULL_BLOCK_TIME_MS is being used. It is the total amount of time to wait, which is set to 2 seconds right now. It means it will wait for up to 2 seconds for all the updates polled in this cycle (only the wait time when inserting into queue). Also it's not typical that all call backs are slow, so as long as one thread is available, they won't be blocked.